### PR TITLE
【main】SP For Qwen3 MoE

### PR DESCRIPTION
**Pull request by lbk-sys**
### What this PR does / why we need it?
Qwen3 MoE supports SP. In scenarios like AlltoAll, AlltoAllv, and MC2, replacing AllReduce with Reduce-Scatter and AllGather achieves computational benefits in norm operations while saving one AllGather communication. This feature is enabled during the P-phase and delivers notable gains in long-sequence scenarios (e.g., 16k–25k), with performance improvements reaching 5%–10%.
### How was this patch tested?
``` 
compilation_config={
    "pass_config":{
        "enable_sequence_parallelism": True
    }
},
enable_expert_parallel=True,
```

- vLLM version: v0.10.0
- vLLM main: https://github.com/vllm-project/vllm/commit/9edd1db02bc6dce6da503503a373657f3466a78b

### CODE REVIEW

File: vllm_ascend/ops/sequence_parallel.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
+import torch
+from torch.nn import functional as F
+from vllm.distributed import (get_tensor_model_parallel_world_size,
+                              get_tp_group, tensor_model_parallel_all_gather,
+                              tensor_model_parallel_reduce_scatter)
+from vllm.forward_context import get_forward_context
+from vllm.platforms import current_platform
```
Comment (wangxiyuan) [inline vllm_ascend/ops/sequence_parallel.py]: Do not use current_platform in vllm-ascend.  import vllm_ascend.platform directlly

File: vllm_ascend/ops/sequence_parallel.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        self.slice_size = self.lengths_sum_padding // self.tp_size
+
+        self.mc2_mask = torch.zeros(
+            self.lengths_sum_padding,
+            dtype=torch.bool,
+            device=current_platform.device_type,
+        )
+        self.mc2_mask[:lengths_sum_unpadding] = True
+
+    def padding_aligned_reduce_scatter(self,
```
Comment (momo609) [inline vllm_ascend/ops/sequence_parallel.py:39]: This functions are duplicated with the pad and unpad functions in flashcommv1,can we aggregating them?

Comment (lbk-sys) [reply vllm_ascend/ops/sequence_parallel.py:39]: Thank you for your suggestion. We chose not to adopt the flashcomm1 implementation from the 091 branch for two reasons:

1. The existing flashcomm1 implementations for Qwen2 and Qwen3 in the repository are inconsistent. We've created a model-level interface here to minimize migration efforts for sparse models (SP).

2. Currently, flashcomm1's graph mode support for sparse models like Qwen2 and Qwen3 isn't available in the main branch. Merging it would impact graph mode performance, so we're keeping it separate for now. Note that merging Qwen3 MoE's SP implementation won't affect the current status.

**Comment by nowmyself-togo-anywhere** [comment]
@lbk-sys Hi, Can you list your vllm serve CLI and your test dataset ? I want to try your pr. Thank you !

**Comment by nowmyself-togo-anywhere** [comment]
@lbk-sys  Hi，can you list the CLI for vllm serve start in your testing scence, and your testing dataset.

**Comment by xueliangyang-oeuler** [comment]
@lbk-sys  can you list out your testing CLI, e.g, vllm serve, and your testing datasets? Thanks.

**Comment by lbk-sys** [comment]
> @lbk-sys can you list out your testing CLI, e.g, vllm serve, and your testing datasets? Thanks.

Thank you for your attention. For testing accuracy, we have run the AIME dataset. For performance testing, we have conducted benchmarks using vLLM and offline scripts. The input data for the model follows the format t*h, so as long as the number of input tokens in the P-phase meets this requirement, benefits can be achieved (regardless of the dataset).

**Comment by lbk-sys** [comment]
> > @lbk-sys can you list out your testing CLI, e.g, vllm serve, and your testing datasets? Thanks.
> 
> Thank you for your attention. For testing accuracy, we have run the AIME dataset. For performance testing, we have conducted benchmarks using vLLM and offline scripts. The input data for the model follows the format t*h, so as long as the number of input tokens in the P-phase meets this requirement, benefits can be achieved (regardless of the dataset).

and deepscaler dataset

File: tests/e2e/multicard/test_offline_inference_distributed.py [hunk 3]
```diff
@@ -234,3 +234,27 @@ def test_models_distributed_DeepSeek_W4A8DYNAMIC():
...
+    ]
+    sampling_params = SamplingParams(max_tokens=5,
+                                     temperature=0.0,
+                                     top_k=50,
+                                     top_p=0.9)
+
+    with VllmRunner(
+            snapshot_download("Qwen/Qwen3-30B-A3B"),
+            dtype="auto",
+            tensor_parallel_size=4,
```
Comment (MengqingCao) [inline tests/e2e/multicard/test_offline_inference_distributed.py]: There is just 2 cards on CI machine, let's reduce tp size to 2

Comment (lbk-sys) [reply tests/e2e/multicard/test_offline_inference_distributed.py]: done ,  thanks

**Review by wangxiyuan — APPROVED** [review]
